### ISSUE TITLE
External Library Design Proposal

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1993]
Preface: This issue evolved as I wrote it. I originally was just planning to summarize my commonality analysis but as I wrote things out I got a pretty clear idea of a design for this, so in the "General thoughts/analysis" section I start laying out those thoughts, then in the "Proposal for the types" section I wrote some code for the ASTs I have in mind. Many of the observations in the commonality analysis are probably too ODE-solver-specific, but hopefully I've done a decent job of generalizing things in the actual analysis/design sections.

---

As a first step towards designing a language for defining external libraries, I'll lay out the results of my commonality analysis for calling the 4 ODE-solving libraries I've been looking at (scipy (Python), Apache commons math (Java), oslo (C#), odeint (C++)).

## Commonality: Complex Input Data

Each library requires defining the input data. In the context of a solver, this is the "thing to solve". In the context of ODE solvers, this is the (system of) ODE(s). By "complex", I mean that the data cannot be expressed with basic types. For example, the data is not just a float or a Boolean.

Variabilities:
- Is the data expressed as a function or a class?
  - If it is a function, 2 cases:
    - parameters are dependent and independent variable (in that order), return a vector/list of the RHSs of the system
    - parameters are dependent, independent, and "result" variable (in some order), function body "sets" the result variable.
  - If a class, contains above function and a constructor that sets the state variables. The state variables are the variables used in the expression of the thing to be solved, other than the dependent and independent variables. The class might also contain more methods, such as `getDimension` for Java.
  - The class may or may not implement an interface

## Commonality: Set up solver object

Each ODE solver I looked at requires defining some sort of solver object of a class provided by the library. For ODE solvers, this is usually referred to as the "integrator". Setting up the integrator always involves specification of the numerical method to use, the absolute tolerance, and the relative tolerance, and sometimes also requires other parameters to pass.

Variabilities:
- The name of the constructor to call. Sometimes the constructor used corresponds to the chosen numerical method, and other times there is a single constructor provided by the library and the numerical method is passed as a parameter.
- The number of methods to call. Sometimes all of the set-up is done in the constructor call, but sometimes you must call additional methods on the object to set up certain properties.
- Are the tolerances passed in as individual parameters, or grouped together in an "Options" record?
- The parameters that need to be passed. Sometimes this set-up of the solver requires passing the time step, initial time, the data representing the thing to solve (ex. the ODE), the initial value, etc.

The variabilities here are extensive enough that it feels like the commonality boils down to "The library provides some constructors and methods for setting up a solver". The details of these are probably too library-specific and therefore cannot be baked into the `ExternalLibrary` data type. In other words, the data type should allow for each value of the type to specify the available methods and their interfaces.

## Commonality: Solve

After the thing to solve and the solver have been defined, using each library then requires performing the actual solving. 

Variabilities:
- Two options:
  - One method call where the initial and final time are parameters, solving the full range at once
  - A method call in a loop, solving one step at a time
- The exact parameters for the call

## Commonality: Initialize solution variable

Using each solver requires defining a variable to hold the results.

If the results are a list, two options:
- Initialize as empty list
- Initialize as list containing the initial value

## Commonality: Populate solution list

This step would not be required if the results are just a single value rather than a list.

Three options, dependent on whether the solving is all at once or one step at a time:
1. The solving is all at once and the population of the solution list happens automatically during solving. In this case, the behaviour for populating this list is encoded in a class, and this class can be considered another instance of "complex data" like what was described in the "complex input data" section above. An object of that class is either:
  - Added to the solver as part of the solver set-up
  - Passed as a parameter to the solve call
2. The solving happens one step at a time in a loop, so in each loop iteration the "current" result is appended to the list. The current result is available on the integrator object.
3. The solving is all at once, and the population of the solution list happens entirely after the solving is complete. In this case, the solve step has stored the results in some undesired format (i.e. an array instead of a list). The list population therefore happens in a loop through the result array/list/collection. 

## General thoughts/analysis

The interfaces to the functions/methods that need to be called are very different between libraries, so the `ExternalLibrary` type should provide a way to specify these interfaces, but cannot make many assumptions about the interfaces to simplify the specification.

This is in contrast to my original thinking, which was that the methods to call would be similar enough that each library would only need to specify the name of the solver constructor, for example, and the parameters would already be known. This would never have generalized to other kinds of libraries anyways, but even when restricting the domain to ODE solving libraries, they are not similar enough for something like that. 

Anyway, my thinking now is that we can represent a library with an ordered list of function descriptions. The order reflects the order that the functions should be called. This gets complicated by the fact that sometimes, based on the numerical method chosen, a method that needs to be called changes. So each item in the list is either mandatory, or is an option between different functions. For each function, we need to specify:
- Is it a function or a method?
  - For methods, need to specify the object it should be called on
- What is it called?
- What are the parameters?
- Is there a return value? Should it be assigned to a variable?

The interesting part is the parameters, and how to specify them. It seems to me that there are at least 2, maybe 3, forms that the parameters can take:
- Complex data, like what is described above in the "Complex input data" section (could split this into 2 different kinds of data, one for functions, one for classes)
  - Need to specify whether it is a class or a function
  - Need to specify the parameters of the class methods, if it is a class (and implements an interface), or the parameters of the function, if it is a function
  - Need to specify whether the class implements an interface
  - Need to specify whether the class is local or top-level (might need to specify this for the function, too)
  - The class or function needs to be defined as part of the code that gets generated (i.e. the generator should know that when complex data is encountered, it needs to generate the corresponding class/function and cannot assume that it already exists)
  - In the case of a class, the declaration/definition of the object also needs to be defined (i.e. the generator should know that it needs to generate the declaration/definition, and cannot assume that the object was already defined previously)
- Simple data
  - Just a value with a regular type
  - Need to specify whether the value needs to be assigned to a variable prior to the function call
- Potential 3rd kind of data for the case of the "Options" record described in the "Set up solver object" section: Record data
  - Specify the fields of the record

To actually use an `ExternalLibrary`, we need to pair it with something that will "fill in the blanks". This data type, let's call it `ExternalLibraryCall` would need to:
- Specify which function was chosen, for the cases where there is an option.
- Specify the values to be used as the parameters to each function or method call
- For complex data, further specify the bodies to use for the functions/methods

Ideally, I'd want to minimize the opportunities for things to go wrong in a pairing. Things could easily go wrong if the `ExternalLibraryCall` doesn't "line up" with the `ExternalLibrary` to which it is being paired (for example, the `ExternalLibraryCall` specifies more or fewer parameters than expected by the specification of the interface for a given function in the `ExternalLibrary`). 

We could prevent all opportunities for things to go wrong if we made `ExternalLibrary` and `ExternalLibraryCall` a single type, instead of using two different types. A value of this type would thus correspond to a single, specific use of the library, and the structure of the type would not allow for things to not "line up". Since we want our external library definitions to be re-usable data, our external library definitions would have to actually be functions, where the signature would be dependent on the number and types of "blanks" that need to be filled in to use that specific library. This seems really weird, perhaps even infeasible, and I don't think we should go in that direction.

So, sticking to the idea of having separate types for `ExternalLibrary` and `ExternalLibraryCall`, we will likely need to rely on users to do things that make sense, and throw errors if they try to pair an `ExternalLibraryCall` to an `ExternalLibrary` that doesn't match up. The structure of the `ExternalLibraryCall` should likely be similar to the structure of the `ExternalLibrary`, but instead of specifying interfaces, we are specifying calls to those interfaces.

An `ExternalLibraryCall` would therefore be an ordered list of function calls. A function call can either be a choice or not (corresponding to a function in `ExternalLibrary` either being a list of options or a mandatory function call). If it's a choice, we need to specify the name of the function we are choosing to call.

We then need to specify a list of values for each call, corresponding to the list of parameters for that function/method in the `ExternalLibrary`. For simple data, this will likely just be an Expr. For complex data, this will include the body of the function or bodies of the class methods, depending on whether the complex data is a function or a class.

## Proposal for the types

At this point, rather than continuing to try to explain my plan for these data types in words, I just started writing some Haskell (pseudo-ish)code for what I think we need. So that is what is below. I've organized it so types are shown in pairs separated by a blank line, the first of each pair being part of the `ExternalLibrary` AST and the second being part of the `ExternalLibraryCall` AST. I thought laying them out this way would help for seeing how the two ASTs resemble each other.

```
type ExternalLibrary = [Step]
type ExternalLibraryCall = [CallStep]

data Step = Choice [FunctionInt] | Mandatory FunctionInt
data CallStep = Chosen FuncName FunctionCall | MandatoryCall FunctionCall

data FunctionInt = F  FuncType FuncName [Parameter] (Maybe Return) Amount
data FunctionCall = CallOnce [ParamValue]  | CallInLoop [ParamValue] [QuantityDict] -- Name not required because ordered list 

data Parameter = Simple VarName CodeType (Maybe VarName) -- Maybe assign result to variable 
  | Fn FuncName [QuantityDict] ([Expr] -> FuncStmt)
  | Class VarName Scope Interface 
  | Record VarName [FieldName]
data ParamValue = SimpleV Expr 
  | FnV [Expr] 
  | ClassV [QuantityDict] MethodInfo  -- QuantityDicts are for the state variables
  | RecordV [Expr]

data Interface = Unknown | Implements String [(FuncName, [QuantityDict])]
data MethodInfo = New [Func] | ImplementWith [[FuncStmt]]

-- Helper types
type VarName = String
type FuncName = String
type Condition = Expr

-- FuncStmt and Func exist in Drasil already, along with functions for translating it to GOOL

data FuncType = Function | Method VarName -- name is of object method should be called on

data Scope = Local | TopLevel

data Return = R VarName CodeType -- Name is variable to assign to

data Amount = Once | Loop ([QuantityDict] -> Condition) ([QuantityDict] -> [FuncStmt]) -- Other statements to do in the loop, need to pass list of QuantityDict for use-case-specific variables
```

I'm sure what I've shown above has problems, but I think it's a good starting point, and I can see what issues arise when I try to actually write the instances of these types for our use cases.

Of course, I'm also hoping for some feedback @JacquesCarette @smiths.

**Comment by JacquesCarette** [comment]
I think it would also be a good idea if you read up on RPC (Remote Procedure Calls); you're not doign RPC, but you are still doing something that requires a 'foreign API'
- https://searchapparchitecture.techtarget.com/definition/Remote-Procedure-Call-RPC
- https://en.wikipedia.org/wiki/Remote_procedure_call
Also relevant is FFI
- https://en.wikipedia.org/wiki/Foreign_function_interface
- https://stackoverflow.com/questions/5440968/understand-foreign-function-interface-ffi-and-language-binding
and its use of IDL
- https://en.wikipedia.org/wiki/Interface_description_language
for specifying calling conventions.  There is some Haskell-specific stuff at https://wiki.haskell.org/Foreign_Function_Interface . Of particular note here is the use of generative techniques for interfaces (under the Tools section)

I'm not saying your pseudo-code is in any way wrong. But you might save yourself some debugging by doing a cross-check with those more general ideas.

**Comment by bmaclach** [comment]
Thanks for the resources @JacquesCarette. After reading them I do have some potential modifications to my pseudo-code. Tagging @smiths as well for his insights, especially about the bolded question below.

The idea of having separate ASTs for 1. Defining the external library and 2. Using the external library is consistent with how FFIs tend to work. They involve first declaring the external functions to use, and then using them in the host language as needed. Using the bindings-dsl library for the Haskell FFI as an example, it provides a macro-based DSL for defining the interface to C functions/types/constants/etc., which then generates a Haskell library that uses the FFI to wrap these functions. Then that Haskell library can be imported into other projects to actually use the C functions it wraps. So my plan seems to fit with this. I have `ExternalLibrary` for defining the interface, then `ExternalLibraryCall` for using the interface.

Where my plan differs from typical FFIs/IDLs is that my interface definition language (`ExternalLibrary`) has baked in a lot of information about the use case for the library. Things like the order in which the functions/methods should be called, whether certain functions should be called once or in a loop, and the variables the return values should be assigned to, are all included in an `ExternalLibrary` definition. I did it this way because, at least for the libraries I looked at, these things seemed like they would stay consistent between uses of the library. There's a certain order in which the functions/methods must be called, independent of the specific ODE being solved or options chosen. I expect this to be true of most "solver" libraries, which I also expect to be the kinds of libraries we are interested in for Drasil (though I'm not certain on this).

So that becomes the question: **Do we expect Drasil to ever want to use a library that might be used in different ways**? The ODE solver libraries are all used in one way -- to solve an ODE. Given an ODE library, the "way" to use it is consistent (the numerical method may change, but this has surprisingly little effect on the overall structure of the library calls). Linear algebra libraries (which in #1986 are mentioned as another kind of library my solution should be able to handle) are used to solve linear systems of equations. Given a linear algebra library, I expect (though I haven't looked closely at these) that again the "way" to use the library will be consistent, no matter the specific system being solved or numerical method chosen. So our current plans lead me to think the answer to the question above might be "no", in which case my pseudo-code is still good as a starting point.

But if the answer to the above question is "yes", or if we want to be more general just to be prepared for that possibility, then I should keep my external library interface definitions free of any use-case specific information. I would propose doing this by splitting my current plan for `ExternalLibrary` into 2 different types, leading to 3 types overall:
1. `ExternalLibrary`, which would encode JUST the external library's API -- the functions/methods it exports, the parameters required by each, and their return type.
2. `ExternalLibraryScenario`, which would build on an `ExternalLibrary` by outlining a specific use-scenario for the library. Basically it would add the use-case specific information that was part of `ExternalLibrary` in my original pseudocode, like the specification of an order in which to call the functions/methods. An external library would only ever have one `ExternalLibrary` definition, but could have many `ExternalLibraryScenario`s defined for it (though our current ODE solvers would probably only have 1 scenario each). 
3. `ExternalLibraryCall`, which wouldn't change much from the `ExternalLibraryCall` definition in my pseudocode above, if at all. An `ExternalLibraryCall` would pair with an `ExternalLibraryScenario` to fill in the remaining missing info, specifically the precise parameter values to use in the various calls. 

One other thing to note is that my plan for defining external library interfaces focused only on the functions/methods exported by the library, whereas the IDLs/FFIs I looked at provided ways of specifying things like constants and types exported as well. I don't see the need for these yet (if a type is defined by the library, it will appear somewhere in the specification of the functions/methods, and thus does not need to be specified separately) but I expect they would be easy extensions to the ASTs I've outlined if we need them in the future (as in, I don't see why adding them would require any sort of major rewrite of the ASTs).

**Comment by JacquesCarette** [comment]
I agree that currently, libraries would be used in a single way.  So I think the 2-types solution is probably still the right one.  I think the hardest thing to handle are callbacks [especially when they involve enumerated types, which is why you might need exported types], and ODE libraries can use callbacks, especially once we start handling events (think of systems with collisions, like simulating a bouncing ball).

**Comment by smiths** [comment]
I agree with @JacquesCarette.  Eventually we might need more layers, but I think we are better served by getting the current task done.  I think we should follow our usual approach of letting the case studies dictate our design, and refactor as new cases are brought in.

**Comment by JacquesCarette** [comment]
Could we get a status update on this?  Should it still be co-assigned to me?

**Comment by bmaclach** [comment]
The work for this is mostly done, just waiting on a PR and merge for the noPCMODE branch (which is waiting on #2014). You don't need to be co-assigned anymore because the questions were answered, I'll remove you from the assignees.

**Comment by JacquesCarette** [comment]
Now that a decision has been made on #2014 what's left here?

**Comment by bmaclach** [comment]
This was closed by #2114.
